linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala

```diff
@@ -1011,28 +1009,32 @@ final class Emitter[E >: Null <: js.Tree](
         }
       }
 
-      _cacheUsed = true
+      _trackerUsed = true
 
-      if (_tree == null || !version.sameVersion(_lastVersion) || (_lastCtor ne ctor) ||
+      if (!version.sameVersion(_lastVersion) || (_lastCtor ne ctor) ||
           !allSame(_lastMemberMethods, memberMethods) ||
           !allSame(_lastExportedMembers, exportedMembers)) {
+        // Input has changed or we were invalidated.
+        // Clean knowledge tracking and re-track dependencies.
         invalidate()
-        _tree = compute
         _lastVersion = version
         _lastCtor = ctor
         _lastMemberMethods = memberMethods
         _lastExportedMembers = exportedMembers
-        (_tree, true)
+
+        (compute, true)
       } else {
-        (_tree, false)
+        // Input has not changed and we were not invalidated.
+        // --> nothing has changed (we recompute to save memory).
+        (compute, false)
```

Comment:
This seems weird. Now, both sides of the `if` call `compute`, and they do not use its result other than to return it. Shouldn't we return only the `Boolean` then, and let the caller do the `compute` themselves? That would avoid the tuple and more importantly the `extractChangedAndWithGlobals`.

Reply:
Hum, yes. You have discovered what I have attempted to brush under the rug :sweat_smile: If I extract `compute` after calling `trackChanged`, `org.scalajs.linker.BasicLinkerBackendTest.noInvalidatedModuleInSecondRun` fails.

I'm still at the stage of debugging where I think "this cannot happen". So all-in-all, I guess this clearly warrants further investigation :P

Reply:
I found the issue: The alternative code was using a short-circuiting or assignment (`||=`) so the RHS was not always executed. Putting it into a `val` first fixed the problem. (updated).